### PR TITLE
Web hosting directly from Peergos

### PR DESCRIPTION
**Pull request by ianopolous**
This allows any Peergos user to publish a website securely without requiring any DNS management or TLS certificates. 

A user publishes a directory, and sets that directory as their webroot in their profile (and makes that profile property public). Then their site is served from http://$username.peergos.localhost:9000 and https://$username.peergos.me

Long term and with browser support this could become a new URL scheme peergos://$username/$path

### CODE REVIEW

File: src/peergos/server/net/GatewayHandler.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            path = path.substring(1).replaceAll("//", "/");
+            if (path.length() == 0)
+                path = "index.html";
+
+            String domain = httpExchange.getRequestHeaders().getFirst("Host");
+            if (! domain.endsWith(domainSuffix))
+                throw new IllegalStateException("Incorrect domain! " + domain);
+            String owner = domain.substring(0, domain.length() - domainSuffix.length());
+            Path toProfileEntry = Paths.get(owner).resolve(".profile").resolve("webroot");
+            AbsoluteCapability capToWebrootFiled = UserContext.getPublicCapability(toProfileEntry, network).join();
```
Comment (kevodwyer) [inline src/peergos/server/net/GatewayHandler.java:43]: spelling of Field

File: src/peergos/server/net/GatewayHandler.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            Optional<FileWrapper> webroot = network.getFile(capToWebroot, owner).join();
+            Optional<FileWrapper> assetOpt = webroot.get().getDescendentByPath(path, crypto.hasher, network).join();
+            if (assetOpt.isEmpty()) {
+                httpExchange.getResponseHeaders().set("Content-Type", "text/plain");
+                httpExchange.sendResponseHeaders(404, 0);
+                return;
+            }
+            FileWrapper file = assetOpt.get();
+            if (file.isDirectory())
+                throw new IllegalStateException("Directory listings not allowed");
```
Comment (kevodwyer) [inline src/peergos/server/net/GatewayHandler.java:57]: HTTP error code in a response?

Comment (ianopolous) [reply src/peergos/server/net/GatewayHandler.java]: I've made it 404 which is what github does

File: src/peergos/server/net/GatewayHandler.java (hunk 1 above)
Comment (ianopolous) [reply src/peergos/server/net/GatewayHandler.java]: fixed

File: src/peergos/server/Main.java [hunk 3]
```diff
@@ -358,7 +358,7 @@ public static final void poststrap(Args args) {
             Stream.of(
                     new Command.Arg("port", "service port", false, "9000"),
                     new Command.Arg("peergos-url", "Address of the Peergos server to connect to", false, "http://localhost:8000"),
-                    new Command.Arg("domain-suffix", "Domain suffix to accept", false, ".public.localhost:9000"),
+                    new Command.Arg("domain-suffix", "Domain suffix to accept", false, ".peergos.localhost:9000"),
```
Comment (kevodwyer) [inline src/peergos/server/Main.java:361]: the comment for the PR references public and not peergos. The PR comment should be updated.

**Review by kevodwyer — APPROVED** [review]

**Pull request merged**